[클린코드 리액트 3기 이진호] 계산기 구현

Gn0lee opened the pull request:
안녕하세요! 온보딩 미션 리뷰 요청 드립니다!

온보딩 미션 진행할때 저의 의도, 고민되었던 점을 간단하게 정리해보았습니다.

1. 컴포넌트 뷰 코드와 기능 코드의 분리

    평소 업무할때 뷰를 구성하는 코드와 기능 또는 비즈니스 로직 코드를 최대한 분리하려고 합니다. 로직과 뷰가 같이 있으면 디자인 변경이나 기능 추가시 유지보수가 매우 어려웠기 때문입니다. 주로 커스텀 훅으로 기능을 분리를 많이 하는 편인 것 같습니다.

2. 계산기 기능 구현 방식
  이번 미션의 계산기는 보통의 계산기와 동작 방식이 약간 달라서 고민을 많이 한 것 같습니다. 처음 계산기 기능을 구현할 때 사칙연산 이외의 연산이 추가 될 때도 유연하게 대응이 되도록 구현하고 싶었습니다. 하지만 이런 기능을 염두하고 구현하려니 경우의 수가 끝이 없어서 우선 사칙연산만 동작하도록 구현하였습니다. 현업에서도 이런 경우가 빈번한 것 같습니다. 나중을 위해 시간을 더 들여서 제대로 만들것인지, 우선 어느정도의 기능을 구현할지에 대한 고민..?? 리뷰어님은 이럴때 어떻게 판단하시는지 궁금합니다.

3. React.StrictMode
  side effect나 상태관리 확인에 도움을 주는 기능이지만 개발 환경에서 코드가 2번실행되는 경우가 꽤 있어서 현재 현업에서는 StrictMode를 사용하지 않고 있습니다. 리뷰어님은 StrictMode를 평소에도 사용하시는지와 사용하신다면 그 이유가 궁금합니다.

부족한 코드이지만 잘 부탁드립니다! 🙇‍♂️

Gn0lee committed:
feat : vite 기반의 react 프로젝트 초기 세팅

Gn0lee committed:
docs : REQUIREMENTS 추가 및 작성

Gn0lee committed:
test : 숫자 입력 테스트 코드 구현

Gn0lee committed:
feat : enter 함수에 숫자 입력 기능 구현

Gn0lee committed:
feat : 숫자 및 연산자 입력 함수 구현

Gn0lee committed:
feat : 연산자 입력 테스트 코드 및 기능 추가

Gn0lee committed:
test : 계산 함수 기능 테스트 코드 추가

Gn0lee committed:
feat : 연산식 계산 수행 기능 구현

Gn0lee committed:
refact : 테스트 중복 코드 모듈화

Gn0lee committed:
test : 초기화 테스트 코드 추가

Gn0lee committed:
feat : 초기화 기능 구현

Gn0lee committed:
test : 음수 계산 테스트 케이스 추가

Gn0lee committed:
feat : 음수 계산 추가

Gn0lee committed:
feat : 이벤트 핸들러 연결

Gn0lee committed:
refact : calculate의 예외처리 분리

excuteOperation 내부에서 alert 발생시키는 것을 calculate 함수 초반에 실행하여 상태 업데이트역할만 수행하도록 수정

yujo11 commented on index.html:
```diff
@@ -0,0 +1,13 @@
+<!doctype html>
+<html lang="en">
```
소소하지만 lang 설정이 en으로 되어 있네요 ㅎㅎ

Gn0lee replied:
평소 신경쓰지 않던 부분이라 lang attribute에 대해 검색을 해봤습니다ㅎㅎ

제가 진행중인 프로젝트의 설정이 어떻게 되어있나 한번 더 살펴봐야 겠네요..ㅎㅎ

yujo11 replied:
lang 설정에 따라 기본으로 적용되는 locale이 변하고, 이에 따라 불필요한 동작들이 발생할 수 있어서 신경 써보시면 좋을거 같아용 ㅎㅎ

yujo11 commented on src/App.tsx:
```diff
@@ -0,0 +1,49 @@
+import useCalculator from 'src/hooks/useCalculator.ts';
+
+function App() {
+	const { enter, calculate, clear, display } = useCalculator();
```
prettier에 Tab 설정이 2 space로 되어있는데 뭔가 4space가 적용된거 같은 느낌이네요

Gn0lee replied:
제 로컬 환경에서는 2space로 되어있고 prettier 실행을 해보아도 더 수정이 안되네요🧐

webstorm 사용중인데 이런 경우가 종종 발생하는지 찾아봐야겠습니다ㅜㅜ

yujo11 commented on src/hooks/useCalculator.test.ts:
오 테스트를 꼼꼼하게 작성해주셨군요 아주 멋지네요 ㅎㅎ 👍

Gn0lee replied:
감사합니다~!🙇‍♂️

yujo11 commented on src/hooks/useCalculator.test.ts:
```diff
@@ -0,0 +1,169 @@
+			expect(result.current.display).toBe('3');
+		});
+
+		it('연산자 입력 후 숫자를 입력하면 해당 숫자가 표시된다.', () => {
```
특정 context에 대한 내용을 한번 더 계층화 시켜줘도 좋을거 같아요

```ts
describe('숫자 입력 테스트', () => {
      describe("연산자가 입력되었을 때", () => {
           test("숫자를 입력하면 해당 숫자가 표시된다.", () => {})
           test("숫자를 3개 초과해서 입력했을 때 에러가 발생한다.", () => {})
       }) 
  })
```

Gn0lee replied:
전혀 생각하지 못했는데 중복 구문을 더 잘 볼수 있는 방법 같네요ㅎㅎ 나중에 또 적용해봐야겠어요

yujo11 commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,102 @@
+			}
+		}
+
+		if (isNaN(result) || !isFinite(result)) {
```
`Number.isNaN`대신 `isNaN`을 사용하신 이유가 있을까요?

Gn0lee replied:
사실 차이점을 모르고 있어서 이번에 검색을 해보았습니다.

isNaN은 숫자로 변환후 Number 판별하고 Number.isNaN은 바로 숫자인지 판별하는 차이점이 있네요. 저는 number의 연산 결과값이 NaN인지 판별해야하니 Number.isNaN이 더 적절한 것 같아서 수정하였습니다ㅎㅎ isFinite도요!

yujo11 replied:
아주 멋지십니다 진호님 ㅎㅎ 👍

yujo11 commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,102 @@
+	const [display, setDisplay] = useState('0');
+
+	const addNumber = (value: string) => {
+		if (display.split(/[+\-x/]/).pop()?.length === 3) {
+			alert('숫자는 3자리까지만 입력 가능합니다.');
```
`3`과 같은 도메인에 연관된 매직넘버들은 상수화가 되어 있으면 의미를 잘 전달할 수 있고, 더 나아가 유지보수에도 이점이 있을거 같네요!

Gn0lee replied:
처음 구현 구상할때 option값을 넘겨줘야겠다 라고 생각만 하고 놓친 부분이네요ㅎㅎ

option에 대해 생각을 하다보니 어디까지 옵션으로 두면 좋을 지 고민을 많이 했던 것 같습니다.

NaN이나 무한대의 경우 동작을 지정하게 해줘야 하나? 같은 고민이 많아져서 일단 구현하고 나중에 생각하자로 넘어갔었습니다.

윤호님은 이럴때 어떻게 판단하시는지 궁금합니다.

yujo11 commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,102 @@
+			const operator = operations[i];
+			const nextValue = parseInt(operations[i + 1]);
+
+			switch (operator) {
+				case '+':
+					result += nextValue;
+					break;
+				case '-':
+					result -= nextValue;
+					break;
+				case 'x':
+					result *= nextValue;
+					break;
+				case '/':
+					if (nextValue === 0) {
+						return '오류';
+					}
+					result = Math.floor(result / nextValue);
+					break;
+				default:
+					return '오류';
+			}
```
이 부분은 Lookup Table을 이용해보시면 보다 간결하게 처리가 가능할거 같네용 ㅎㅎ 


```ts
// ex)

const calculator = {
  "+": ~~
  "-": ~~
  ...
}

calculator?.[operator] ?? 에러처리로직
```

Gn0lee replied:
switch 구문을 자주 사용했었는데 Lookup table를 처음 알게 되었네요ㅎㅎ

확실이 코드가 간결하고 읽기 더 편한 것 같아요

yujo11 commented on src/hooks/useCalculator.ts:
훅으로 비지니스 로직을 깔끔하게 분리하신 부분 아주 좋네용 ㅎㅎ 👍

yujo11 commented on src/hooks/useCalculator.ts:
정규표현식은 굉장히 편리하지만 가독성이 떨어져서 유지보수 측면에서 좋지 않은 부분도 있다고 생각하는데요. 그렇기 때문에 정규표현식을 사용하지 않고 코드로 표현할 수 있는 부분이 최대한 코드로 표현되면 좋다고 생각합니다.

진호님은 이 부분에 대해 어떻게 생각하시는지 궁금하네용 ㅎㅎ

Gn0lee replied:
저는 정규표현식으로 판별하는 것이 javascript코드로 구현하는 것보다 간결한 부분이 있어서 정규 표현식을 선호합니다.

하지만 코멘트를 보고 생각해보니 정규표현식을 이해하는데에도 지식과 노력이 필요해 가독성이 떨어지는 것이 맞는 것 같네요.

그래서 정규표현식을 그대로 사용하기 보다 상수로 관리해서 의도를 알기 쉽게 하면 좋다고 생각하여 상수로 리팩터링 해보았습니다.

yujo11 commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,102 @@
+		throw new Error('잘못된 입력입니다.');
+	};
+
+	const executeOperation = (prevState: string) => {
```
이 함수 하나에서 뭔가 많은 일들이 일어나고 있는거 같은데 적절한 단위로 코드를 분리하면 가독성이 좋아지지 않을까하는 생각이 드네요 ㅎㅎ

Gn0lee replied:
처음 작성할때 for 반복문을 사용하는 것 부터가 가독성이 떨어진다고 생각했는데 방법이 떠오르지 않아서 일단 for 반복문으로 구현하였습니다.

위의 lookup table 작업을 하다보니 tuple 형식이면 reduce로 처리할 수 있겠다 생각이 들어서 방향을 수정해보았습니다.

tuple을 얻는 과정, tuple을 가지고 연산을 수행하는 과정으로 분리하는 리팩터링을 진행해보았습니다

yujo11 commented on src/App.tsx:
onClick의 callback이 모두 익명함수로 전달되고 있는데 익명함수를 사용하시는 진호님의 기준이 궁금합니다!

Gn0lee replied:
보통 onClick의 동작이 매우 간단하거나 map 함수 안에 있는 경우 익명 함수를 사용합니다.

App 컴포넌트를 다시 보니 익명 함수로 되어있으니 return 부분이 너무 길어져서  가독성이 떨어지는 것 같기도 해서 다른 기준을 삼아야 하나 고민이 되네요.

혹시 윤호님의 callback 전달 기준을 여쭤봐도 될까요??

yujo11 replied:
익명함수를 넘기게 되면 렌더링할 때마다 함수를 재생성해서 성능과 연관 된 부분도 있고, 익명함수로는 의미 전달이 쉽지 않은 부분도 있기 떄문에 저는 특별한 경우를 제외하고는 기명함수를 많이 사용하는 편인거 같아요 ㅎㅎ

yujo11 commented on src/App.tsx:
```diff
@@ -0,0 +1,49 @@
+			<div className="modifiers subgrid">
+				<button
+					className="modifier"
+					onClick={() => {
+						clear();
+					}}
```
아래처럼만 작성해도 동작할거 같은데 익명함수여야 하는 이유가 있었는지 궁금하네용 ㅎㅎ

```suggestion
					onClick={clear}
```

Gn0lee replied:
처음 커스텀 훅을 구현할때 "컴포넌트에서 전달받은 함수들(enter, caculate, clear)을 실행한다" 라고 생각하였습니다. 그래서 AC 클릭시 clear를 실행시키기 위해 익명함수로 구현하였습니다.

클릭 말고도 다른 경우에도 clear를 실행시킬 수 있다고 생각했기 때문입니다.

다시 생각해보니 가독성 측면에서 익명함수를 사용하지 않는 것이 더 좋을 것 같긴하네요.

다음 미션때는 이런 부분을 더 생각해보면서 구현해봐야 할 것 같아요🥲

yujo11 reviewed (changes requested):
안녕하세요 진호님 ~

계산기 미션 함께하게된 조윤호입니다. 잘 부탁드립니다 ㅎㅎ

비지니스 로직을 훅으로 분리한 부분이 좋았고 전체적으로 코드의 흐름을 따라가면서 이해하기 어려운 부분이 크게 없었네요 👍 


>평소 업무할때 뷰를 구성하는 코드와 기능 또는 비즈니스 로직 코드를 최대한 분리하려고 합니다. 로직과 뷰가 같이 있으면 디자인 변경이나 기능 추가시 유지보수가 매우 어려웠기 때문입니다. 주로 커스텀 훅으로 기능을 분리를 많이 하는 편인 것 같습니다.

좋은 습관인거 같습니다 ㅎㅎ 조금 더 나아가면 코드에 대한 테스트를 작성할 때도 자주 변하는 로직(View)과 자주 변하지 않는 로직(Domain)을 분리해두면 테스트 작성하기도 훨씬 수월해지더라구요. ㅎㅎ

>이번 미션의 계산기는 보통의 계산기와 동작 방식이 약간 달라서 고민을 많이 한 것 같습니다. 처음 계산기 기능을 구현할 때 사칙연산 이외의 연산이 추가 될 때도 유연하게 대응이 되도록 구현하고 싶었습니다. 하지만 이런 기능을 염두하고 구현하려니 경우의 수가 끝이 없어서 우선 사칙연산만 동작하도록 구현하였습니다. 현업에서도 이런 경우가 빈번한 것 같습니다. 나중을 위해 시간을 더 들여서 제대로 만들것인지, 우선 어느정도의 기능을 구현할지에 대한 고민..?? 리뷰어님은 이럴때 어떻게 판단하시는지 궁금합니다.

저는 [YAGNI](https://ko.wikipedia.org/wiki/YAGNI#:~:text=YAGNI(You%20aren't%20gonna,(XP)%EC%9D%98%20%EC%9B%90%EC%B9%99%EC%9D%B4%EB%8B%A4.) 원칙을 지키려고 하는 편인거 같아요 ㅎㅎ

처음부터 많은 부분을 추상화하면 후에 있을 변경에 유연하게 대응할 수 있는 부분도 존재하겠지만 추상화레벨이 높아짐에 따라 코드를 이해하기 어려워지는 측면도 있고, 미래에 대해서는 알 수 없기 때문에 현재 시점에서 불필요한 코드를 추가로 많이 붙이려고 하지는 않는거 같습니다

또한 기본적으로 역할과 책임을 분명히 나누고 결합도는 낮추고, 응집도를 높여서  코드를 작성하면 후에 변경에 충분히 유연하게 대응이 가능하다고 생각합니다!

>React.StrictMode
side effect나 상태관리 확인에 도움을 주는 기능이지만 개발 환경에서 코드가 2번실행되는 경우가 꽤 있어서 현재 현업에서는 StrictMode를 사용하지 않고 있습니다. 리뷰어님은 StrictMode를 평소에도 사용하시는지와 사용하신다면 그 이유가 궁금합니다.

저는 StrictMode를 사용하고 있고, 사용을 선호하는 편입니다. StrictMode에서 부적절한 API사용과 불필요한 리렌더링을 감지해주기 때문인데요. 이런 부분들은 하나씩 발견되면 수정에 큰 어려움이 없지만 후에 한번에 여러군데에서 발견하게 된다면 수정이 어려워질 수 있는 부분이고, dev모드에서만 실행되어 실수를 방지해주기 때문에 개인적으로는 유용한 기능이라고 생각하고 있습니다 ㅎㅎ

---

업무 진행하시면서 미션 진행하는게 쉽지 않으셨을텐데 고생 많으셨습니다 진호님!

미션이 밀리게 되면 매주 이뤄지는 피드백 강의에 집중하지 못할 수 있기 때문에 이어지는 다른 미션들도 기한 내에 완수하는걸 최우선 목표로 진행해보시면 더 좋겠습니다!

코멘트 확인하고 다시 리뷰요청 부탁드립니다~~! 👍

Gn0lee committed:
refact : index.html lang 속성 수정

Gn0lee committed:
refact : 테스트 조건이 겹치는 테스트들 그룹화

- 마지막 입력이 연산자인 경우 테스트를 그룹화

Gn0lee committed:
refact : NaN, finite 함수 Number 객체 메소드로 수정

Gn0lee committed:
refact : 입력 가능 숫자 길이 props로 수정

Gn0lee committed:
refact : executeOperation 역할 분리 및 lookup table 적용

Gn0lee committed:
refact : regex 상수화 및 excuteOperation try catch 제거

- REGEX의 의도를 알기 쉽도록 변수명에 설명
- lookup table에 해당하지 않는 변수에러는 작업자의 실수로 판단하여 외부 코드에서 핸들링하는 것이 적절하다 판단

Gn0lee commented:
Lookup table, isNaN과 Number.isNaN의 차이점, 익명함수에대한 생각 등 새로 알게 된 지식도 있고 제 코딩 습관도 다시 생각해 볼 수 있어서 너무 좋았습니다.

그리고 질문에 대한 답변도 감사합니다. YAGNI라는 말을 처음 알게 되어서 평소 저의 관점을 다시 생각해 본 것 같아요ㅎㅎ 그리고 미션중에는 StrictMode를 사용해볼 계획이에요ㅎㅎ

좋은 리뷰 감사합니다!

yujo11 commented on index.html:
```diff
@@ -1,5 +1,5 @@
 <!doctype html>
-<html lang="en">
+<html lang="ko">
```
👍 👍 👍

yujo11 commented on src/hooks/useCalculator.test.ts:
```diff
@@ -44,22 +44,24 @@ describe('useCalculator 커스텀 훅 테스트', () => {
+				enterOperators(['+', '1', '2', '3', '4']);
+
+				expect(global.alert).toHaveBeenCalledWith('숫자는 3자리까지만 입력 가능합니다.');
+			});
```
오 테스트가 훨씬 명확해졌네요! 👍 👍 👍 💯 💯

yujo11 commented on src/hooks/useCalculator.ts:
```diff
@@ -78,7 +78,7 @@ export default function useCalculator() {
 		}
 
-		if (isNaN(result) || !isFinite(result)) {
+		if (Number.isNaN(result) || !Number.isFinite(result)) {
```
오 아주 좋네용 ㅎㅎ 👍

yujo11 commented on src/hooks/useCalculator.ts:
```diff
@@ -4,6 +4,14 @@ interface CalculatorOptions {
 	maxDigits?: number;
 }
 
+const REGEX = Object.freeze({
+	IS_NUMBER: /^[0-9]$/,
+	IS_OPERATOR: /^[+\-x/]$/,
+	OPERATOR_SPLITTER: /([+\-x/])/g,
+	IS_ENDS_WITH_NUMBER: /\d$/,
+	IS_ENDS_WITH_OPERATOR: /[+\-x/]$/,
+});
```
이전보다 코드 보기가 편해졌네요! 👍

yujo11 reviewed (approved):
안녕하세요 진호님~

빠르게 코멘트 확인하고 반영해주셨군요 ㅎㅎ 👍 

제가 남긴 코멘트가 도움이 되었다니 다행입니다 ㅎㅎ 추가로 남겨주신 질문들에도 코멘트 남겨놨으니 확인 부탁드려요!

이번 계산기 미션은 여기서 이만 merge해도 좋을거 같아요! 다음 미션도 화이팅입니다 진호님~~ 👍 👍 👍
